src/key/helper.js

```diff
@@ -100,41 +100,88 @@ export async function createBindingSignature(subkey, primaryKey, options, config
   return subkeySignaturePacket;
 }
 
 /**
- * Returns the preferred signature hash algorithm of a key
- * @param {Key} [key] - The key to get preferences from
- * @param {SecretKeyPacket|SecretSubkeyPacket} keyPacket - key packet used for signing
+ * Returns the preferred signature hash algorithm for a set of keys.
+ * @param {Array<Key>} [targetKeys] - The keys to get preferences from
+ * @param {SecretKeyPacket|SecretSubkeyPacket} signingKeyPacket - key packet used for signing
  * @param {Date} [date] - Use the given date for verification instead of the current time
- * @param {Object} [userID] - User ID
+ * @param {Object} [targetUserID] - User IDs corresponding to `targetKeys` to get preferences from
  * @param {Object} config - full configuration
  * @returns {Promise<enums.hash>}
  * @async
  */
-export async function getPreferredHashAlgo(key, keyPacket, date = new Date(), userID = {}, config) {
-  let hashAlgo = config.preferredHashAlgorithm;
-  let prefAlgo = hashAlgo;
-  if (key) {
-    const selfCertification = await key.getPrimarySelfSignature(date, userID, config);
-    if (selfCertification.preferredHashAlgorithms) {
-      [prefAlgo] = selfCertification.preferredHashAlgorithms;
-      hashAlgo = crypto.hash.getHashByteLength(hashAlgo) <= crypto.hash.getHashByteLength(prefAlgo) ?
-        prefAlgo : hashAlgo;
+export async function getPreferredHashAlgo(targetKeys, signingKeyPacket, date = new Date(), targetUserIDs = [], config) {
+  /**
+   * If `preferredSenderAlgo` appears in the prefs of all recipients, we pick it; otherwise, we use the
+   * strongest supported algo (`defaultAlgo` is always implicitly supported by all keys).
+   * if no keys are available, `preferredSenderAlgo` is returned.
+   * For ECC signing key, the curve preferred hash is taken into account as well (see logic below).
+   */
+  const defaultAlgo = enums.hash.sha256; // MUST implement
+  const preferredSenderAlgo = config.preferredHashAlgorithm;
+
+  const supportedAlgosPerTarget = await Promise.all(targetKeys.map(async (key, i) => {
+    const selfCertification = await key.getPrimarySelfSignature(date, targetUserIDs[i], config);
+    const targetPrefs = selfCertification.preferredHashAlgorithms;
+    return targetPrefs;
+  }));
+  const supportedAlgosMap = new Map(); // use Map over object to preserve numeric keys
+  for (const supportedAlgos of supportedAlgosPerTarget) {
+    for (const algo of supportedAlgos) {
+      supportedAlgosMap.set(algo, supportedAlgosMap.has(algo) ? supportedAlgosMap.get(algo) + 1 : 1);
     }
   }
-  switch (keyPacket.algorithm) {
-    case enums.publicKey.ecdsa:
-    case enums.publicKey.eddsaLegacy:
-    case enums.publicKey.ed25519:
-    case enums.publicKey.ed448:
-      prefAlgo = crypto.getPreferredCurveHashAlgo(keyPacket.algorithm, keyPacket.publicParams.oid);
+  const isSupportedHashAlgo = hashAlgo => targetKeys.length === 0 || supportedAlgosMap.get(hashAlgo) === targetKeys.length || hashAlgo === defaultAlgo;
```

Comment:
I think we also need some logic to handle the case where all recipients support a hash algorithm that we don't support? Otherwise `crypto.hash.getHashByteLength` below will throw, I think?